### PR TITLE
create a subkernel registration system

### LINKED ISSUE #4: Create registration system
In the style of papermill registration, create a system that allows us to register how we go from config text to kernel manager options. See `papermill_io.register` as prime examples `papermill_translators.register`.

The initial idea is we create our own magics registration namespace. It would look something like this for a user of the kernel:

```
%%kernel.conda-env-yaml
name: env-name
channels:
  - conda-forge
  - defaults
dependencies:
  - python=3.7
  - codecov
```

Whereas the implementer has their own package that calls our registration:

```python
pick_kernel.kernels.register("conda-env-yaml", handler)
```

What do you think @MSeal & @willingc?

### PR DESCRIPTION
**Pull request by rgbkrk**
Closes #4.

This creates a basic registration system that allows setting up new subkernels. The current API is roughly this:

```python
from pick_kernel.subkernel import Subkernel, pick_subkernels

class MyKernel(Subkernel):
    async def launch(config, session, context, connection_file):
        return kernel_manager_that_has_already_started_a_kernel

pick_subkernels.register("wowsers", MyKernel)
```

Which would then be usable as:

```python
%%kernel.wowsers

config: totally
```

### CODE REVIEW

**Review by MSeal — APPROVED** [review]
This looks nice. Do you need any help adding entrypoint registration (not needed for first pass imo)?

**Comment by rgbkrk** [comment]
> Do you need any help adding entrypoint registration (not needed for first pass imo)?

I'll let you know if I have any questions after reading how the papermill code does it.

Before I merge this, I'm going to do another pass on cleaning up some of this API.

**Comment by rgbkrk** [comment]
It turned out simpler than I expected to create a conda-env-by-name kernel magic! Here's how to do a _rough_ conda kernel implementation:

```python
import os
import sys
from shlex import quote

class AsyncCondaKernelManager(AsyncKernelManager):
    def __init__(self, *args, conda_root=None, conda_environment=None, **kwargs):
        super(AsyncCondaKernelManager, self).__init__(*args, **kwargs)

        self.conda_root = conda_root
        self.conda_environment = conda_environment

    def format_kernel_cmd(self, extra_arguments=None):
        raw_command = ["python", "-m", "ipykernel_launcher", "-f", self.connection_file]
        command = " ".join(quote(c) for c in raw_command)

        activate = os.path.join(self.conda_root, "bin", "activate")

        script = ". '{}' '{}' && exec {}".format(
            activate, self.conda_environment, command
        )

        kernel_cmd = ["bash", "-c", script]

        return kernel_cmd


class CondaKernel(Subkernel):
    @staticmethod
    async def launch(config, session, context, connection_file):
        conda_env = config.split("\n")[0]

        km = AsyncCondaKernelManager(
            # TODO: make this configurable or just detect it...
            conda_root="/Users/kylek/opt/miniconda3/",
            conda_environment=conda_env,
            kernel_name="python3",
            # Pass our IPython session as the session for the KernelManager
            session=session,
            # Use the same ZeroMQ context that allows for awaiting on recv
            context=context,
            connection_file=connection_file,
        )

        await km.start_kernel()

        return km

_subkernels.register("conda", CondaKernel)
```

Then the user runs:

```python
%%kernel.conda
example-environment
```

It's delightful to see this work:

image

**Comment by MSeal** [comment]
That is a satisfyingly low amount of code to handle that complexity

**Comment by rgbkrk** [comment]
The main thing I'd be worried about with this conda kernel implementation is if it's ok that `bash -c` is the top-level process. If we later need to send a signal to the process, will it get passed to the kernel itself (worth noting that interrupts are not currently passed from the pick kernel to the child kernel, so I haven't tested this).

**Comment by rgbkrk** [comment]
I suppose the next question is whether or not we include this conda kernel. I wrote it mostly to see what our interface looks like here and if it's worth keeping it as is.

**Comment by MSeal** [comment]
> Because of the way the process tree works under Unix, if the process created by Popen spawns sub-processes, those children will not receive any signals sent to the parent. That means, for example, it will be difficult to cause them to terminate by sending SIGINT or SIGTERM.

I think it should be ok with an exec because the child process replaces the bash process. If there's still issues https://stackoverflow.com/questions/4789837/how-to-terminate-a-python-subprocess-launched-with-shell-true has some patterns to apply to work around the process signalling

**Comment by MSeal** [comment]
> I suppose the next question is whether or not we include this conda kernel. I wrote it mostly to see what our interface looks like here and if it's worth keeping it as is.

Hmmm it might be best not in an examples folder that docs reference for how to use? The hardcoded env path seems problematic for shared notebooks across systems.

**Comment by rgbkrk** [comment]
> The hardcoded env path seems problematic for shared notebooks across systems.

Haha, yes. I was going to set that up as config options to the `CondaKernel`, not stick my path in the repo. 😆 

Examples folder sounds good.

**Comment by rgbkrk** [comment]
I've been working on handling of errors due to kernels not being registered, but I think I'll just merge this and make some new PRs.

**Pull request merged**
